### PR TITLE
feat(Vizzy): new presence

### LINKED ISSUE #6030: Vizzy | vizzy.io
### Website name

Vizzy.io

### Website URL

https://vizzy.io/

### Prerequisites

- [ ] It is a paid service
- [ ] It displays NSFW content
- [ ] It is region restricted

### Description

Create a professional music video in our free powerful Online Editor.

### PR DESCRIPTION
**Pull request by BigBoi-TM**
## Description 

Resolves #6030
Features:

- Detects homepage, community page, user page, editor, export page, etc.
   - Editor: detects when editing owned projects or community projects
   - Exporter: displays exporting status (I.e.: waiting, estimated time left, etc.)

`*Note: most elements are found using xPath because they have random id's`
## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

### - Homepage
image

### - Community
`P.S.: Same page goes to two paths`
image
image

### - Project
image

### - User
image

### - Editor
`v Community Project v`
image
`v Owned Project v`
image

### - Exporting
`v Awaiting Export v`
image
`v Exporting v`
image
image

</details>

### CODE REVIEW

File: websites/V/Vizzy/presence.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		null
+	).singleNodeValue;
+}
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+			smallImageKey: "logo_512",
+			startTimestamp: browsingTimestamp
+		},
+		{ pathname, search } = window.location;
+	switch (true) {
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:20]: all you do in this switch statement is compare a variable to a string so please do a proper switch statement with the variable in here and the cases down below

File: websites/V/Vizzy/presence.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+			break;
+		}
+		case pathname === "/project": {
+			presenceData.largeImageKey = "discover";
+			presenceData.details = "Viewing a project";
+			presenceData.state = `Project: ${
+				getElementByXpath(
+					"/html/body/div[1]/div/main/div/div/div[2]/div/div[2]/div[1]/h4"
+				).textContent
+			}`;
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:38]:
```suggestion
			presenceData.state =
				getElementByXpath(
					"/html/body/div[1]/div/main/div/div/div[2]/div/div[2]/div[1]/h4"
				).textContent;
```

File: websites/V/Vizzy/presence.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+					label: "View User",
+					url: location.href
+				}
+			];
+			break;
+		}
+		case pathname === "/profile": {
+			presenceData.details = "Viewing their profile";
+			presenceData.largeImageKey = "profile";
+			if (search) presenceData.state = `Tab: ${search.split("=")[1]}`;
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:64]: this isn't reliable, you should use URLSearchParams or a regex

File: websites/V/Vizzy/presence.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+			presenceData.largeImageKey = "export";
+			if (getElementByXpath("/html/body/div[1]/div/div/div[2]/div[2]/p")) {
+				presenceData.details = "Exporting a video";
+				presenceData.state = getElementByXpath(
+					"/html/body/div[1]/div/div/div[2]/div[2]/p"
+				).textContent;
+			} else {
+				presenceData.details = "Preparing to export a video";
+				/* if (document.querySelector<HTMLDivElement>('id="notistack-snackbar'))
+					presenceData.state = "Status: Export Complete..."; */
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:94]: remove

File: websites/V/Vizzy/presence.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+				presenceData.state = "Status: Waiting...";
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:95]: do you really need this here?

File: websites/V/Vizzy/presence.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+			}
+			break;
+		}
+		case pathname === "/discover": {
+			if (search === "?tab=faq") {
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:100]: you should use URLSearchParams

File: websites/V/Vizzy/presence.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+				presenceData.details = "Reading frequently asked questions";
+				presenceData.largeImageKey = "faq";
+			}
+			break;
+		}
+		case pathname === "/privacy": {
+			presenceData.largeImageKey = "faq";
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:107]: you sure you want to put the faq image key on the privacy page?

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/V/Vizzy/presence.ts (hunk 2 above)
Comment (BigBoi-TM) [reply websites/V/Vizzy/presence.ts:38]: Can't keep the `Project:` part?

Comment (ImRodry) [reply websites/V/Vizzy/presence.ts:38]: No, you already mention the user is viewing a project in the presence details right above

File: websites/V/Vizzy/presence.ts [hunk 8]
```diff
@@ -0,0 +1,129 @@
+const presence = new Presence({
+		clientId: "968353669491871754"
+	}),
+	browsingTimestamp = Math.floor(Date.now());
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:4]: what are you flooring? An integer?

File: websites/V/Vizzy/presence.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+		case "/project": {
+			presenceData.largeImageKey = "discover";
+			presenceData.details = "Viewing a project";
+			presenceData.state = getElementByXpath(
+				"/html/body/div[1]/div/main/div/div/div[2]/div/div[2]/div[1]/h4"
+			).textContent;
+			presenceData.buttons = [
+				{
+					label: "View Project",
+					url: location.href
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:41]: Declare href in the destructure at line 19
```suggestion
					url: href
```

File: websites/V/Vizzy/presence.ts [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+		case "/user": {
+			presenceData.largeImageKey = "discover";
+			presenceData.details = "Viewing a user's page";
+			presenceData.state = `@${
+				getElementByXpath("/html/body/div[1]/div/h5").textContent
+			}`;
+			presenceData.buttons = [
+				{
+					label: "View User",
+					url: location.href
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:55]:
```suggestion
					url: href
```

File: websites/V/Vizzy/presence.ts [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+		XPathResult.FIRST_ORDERED_NODE_TYPE,
+		null
+	).singleNodeValue;
+}
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+			smallImageKey: "logo_512",
+			startTimestamp: browsingTimestamp
+		},
+		{ pathname, search } = window.location;
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:19]:
```suggestion
		{ pathname, search, href } = window.location;
```

File: websites/V/Vizzy/presence.ts [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+				}
+			];
+			break;
+		}
+		case "/profile": {
+			presenceData.details = "Viewing their profile";
+			presenceData.largeImageKey = "profile";
+			if (search) {
+				presenceData.state = `Tab: ${new URLSearchParams(
+					document.location.search
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:65]: make some sense please
```suggestion
					search
```

File: websites/V/Vizzy/presence.ts [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+			if (getElementByXpath("/html/body/div[1]/div/div/div[2]/div[2]/p")) {
+				presenceData.details = "Exporting a video";
+				presenceData.state = getElementByXpath(
+					"/html/body/div[1]/div/div/div[2]/div[2]/p"
+				).textContent;
+			} else presenceData.details = "Preparing to export a video";
+			break;
+		}
+		case "/discover": {
+			if (new URLSearchParams(document.location.search).get("tab") === "faq") {
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:97]: just declare the searchparams variable at the top

File: websites/V/Vizzy/presence.ts [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+				new URLSearchParams(document.location.search).get("tab") === "lyrics"
+			) {
+				presenceData.largeImageKey = "faq";
+				presenceData.details = "Viewing an unsupported page";
+			} else if (!search) presenceData.details = "Browsing features";
+			break;
+		}
+		case "/privacy": {
+			presenceData.largeImageKey = "discover";
+			presenceData.details = "Reading privacy terms";
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:115]:
```suggestion
			presenceData.details = "Reading the privacy policy";
```

File: websites/V/Vizzy/presence.ts [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+			break;
+		}
+		case "/terms-of-service": {
+			presenceData.largeImageKey = "discover";
+			presenceData.details = "Reading terms of use";
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:120]:
```suggestion
			presenceData.details = "Reading the terms of use";
```

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/V/Vizzy/presence.ts [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+			presenceData.largeImageKey = "editor";
+			break;
+		}
+		case "/export": {
+			presenceData.largeImageKey = "export";
+			if (getElementByXpath("/html/body/div[1]/div/div/div[2]/div[2]/p")) {
+				presenceData.details = "Exporting a video";
+				presenceData.state = getElementByXpath(
+					"/html/body/div[1]/div/div/div[2]/div[2]/p"
+				).textContent;
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:92]: should probably only compute this once

File: websites/V/Vizzy/presence.ts [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+			} else presenceData.details = "Preparing to export a video";
+			break;
+		}
+		case "/discover": {
+			const tabParams = new URLSearchParams(search).get("tab");
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:97]: I believe I suggested making a searchParams variable at the top so you don't have to compute this so often

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/V/Vizzy/presence.ts (hunk 12 above)
Comment (BigBoi-TM) [reply websites/V/Vizzy/presence.ts]: Cuz it was addressed; replaced with the search variable `tabParams`

File: websites/V/Vizzy/presence.ts [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+					label: "View User",
+					url: href
+				}
+			];
+			break;
+		}
+		case "/profile": {
+			presenceData.details = "Viewing their profile";
+			presenceData.largeImageKey = "profile";
+			if (search) presenceData.state = `Tab: ${tabParams}`;
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:64]:
```suggestion
			if (tabParams) presenceData.state = `Tab: ${tabParams}`;
```

File: websites/V/Vizzy/presence.ts [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+			presenceData.largeImageKey = "editor";
+			break;
+		}
+		case "/export": {
+			presenceData.largeImageKey = "export";
+			presenceData.details = "Exporting a video";
+			break;
+		}
+		case "/discover": {
+			// eslint-disable-next-line unicorn/prefer-switch
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:90]: uhm no? Don't disable it

**Review by ImRodry — CHANGES_REQUESTED** [review]

**Comment by ImRodry** [comment]
Rebased to apply the latest changes in the prettier configuration. I also squashed all the commits in your branch as the commit history was quite messy. If you need help with your local branch, please ping me on the #presence-dev channel on Discord.

**Comment by BigBoi-TM** [comment]
bruh

**Comment by BigBoi-TM** [comment]
finally 😤

File: websites/V/Vizzy/presence.ts [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+					presenceData.details = "Reading about creator mode";
+					break;
+				}
+				case "lyrics": {
+					presenceData.smallImageKey = "faq";
+					presenceData.details = "Viewing an unsupported page";
+					break;
+				}
+				default:
+					if (!search) presenceData.details = "Browsing features";
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:107]: what if there is search and there isn't a tab param in search? Wouldn't you want to display the same thing?

**Review by ImRodry — CHANGES_REQUESTED** [review]

**Review by EncryptedDev — APPROVED** [review]

File: websites/V/Vizzy/presence.ts [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+					presenceData.details = "Reading about creator mode";
+					break;
+				}
+				case "lyrics": {
+					presenceData.smallImageKey = "faq";
+					presenceData.details = "Viewing an unsupported page";
+					break;
+				}
+				default:
+					if ((search && !tabParams) || !search)
```
Comment (ImRodry) [inline websites/V/Vizzy/presence.ts:107]: Works the same
```suggestion
					if (!tabParams || !search)
```

**Review by ImRodry — APPROVED** [review]

**Comment by ImRodry** [comment]
what even happened here

**Comment by BigBoi-TM** [comment]
Don't look at me, I ain't touch nothing

**Comment by ImRodry** [comment]
there are no prettier issues, the CI is definitely not ok, we'll have to fix this

**Pull request merged**